estimagic/decorators.py

```diff
@@ -0,0 +1,140 @@
+                        params["_post_replacements"].to_numpy().astype(int)
+                    ),
+                    processed_params=params,
+                )
+
+            out = func(p, *args, **kwargs)
+
+            return out
+
+        return wrapper_x_to_params
+
+    return decorator_x_to_params
+
+
+def log_parameters_and_criterion_value(database, tables):
+    """Log parameters and fitness values."""
+
+    def decorator_log_parameters_and_criterion_value(func):
+        @functools.wraps(func)
+        def wrapper_log_parameters_and_criterion_value(params, *args, **kwargs):
+            criterion_value = func(params, *args, **kwargs)
+
+            if database:
+                adj_params = params.copy().set_index("name")["value"]
+                append_rows(database, tables, [adj_params, {"value": criterion_value}])
+
+            return criterion_value
+
+        return wrapper_log_parameters_and_criterion_value
+
+    return decorator_log_parameters_and_criterion_value
+
+
+def log_gradient(database, names):
+    """Log gradient."""
+
+    def decorator_log_gradient(func):
+        @functools.wraps(func)
+        def wrapper_log_gradient(*args, **kwargs):
+            gradient = func(*args, **kwargs)
+
+            if database:
+                append_rows(
+                    database, ["gradient_history"], [dict(zip(names, gradient))]
+                )
+
+            return gradient
+
+        return wrapper_log_gradient
+
+    return decorator_log_gradient
+
+
+def log_gradient_status(database, n_gradient_evaluations):
+    """Log the gradient status.
+
+    The gradient status is between 0 and 1 and shows the share of function evaluations
+    to compute the gradients.
+
+    """
+    counter = itertools.count(1)
+
+    def decorator_log_gradient_status(func):
+        @functools.wraps(func)
+        def wrapper_log_gradient_status(params, *args, **kwargs):
+            criterion_value = func(params, *args, **kwargs)
+
+            if database:
+                c = next(counter)
+                if n_gradient_evaluations is None:
+                    status = c
+                else:
+                    status = (c % n_gradient_evaluations) / n_gradient_evaluations
+                    status = 1 if status == 0 else status
+                update_scalar_field(database, "gradient_status", status)
+
+            return criterion_value
+
+        return wrapper_log_gradient_status
+
+    return decorator_log_gradient_status
+
+
+def exception_handling(start_params, general_options):
+    def decorator_exception_handling(func):
+        @functools.wraps(func)
+        def wrapper_exception_handling(x, *args, **kwargs):
+            try:
+                out = func(x, *args, **kwargs)
+            except (KeyboardInterrupt, SystemExit):
+                raise
+            except Exception as e:
+                constant, slope = general_options.get(
+                    "criterion_exception_penalty", (np.nan, np.nan)
```

Comment:
I would prefer to set the default penalty constant to something like 2 * criterion(start_params) which we can save in the copy of general_options after the initial function evaluation. The default slope could be `0.1 * criterion(start_params)` or some other relatively small fraction of the start function value. 

I think reasonable defaults that do not depend on the order of magnitude of the initial function evaluation are impossible. I also think that the slope should not be zero by default (which it currently is) because users would never think about switching it on.